src/bootstrap/src/core/config/config.rs

```diff
@@ -1325,23 +1325,14 @@ impl Config {
             .into_iter()
             .chain(flags.exclude)
             .map(|p| {
-                let p = if cfg!(windows) {
+                // Never return top-level path here as it would break `--skip`
+                // logic on rustc's internal test framework which is utilized
+                // by compiletest.
+                if cfg!(windows) {
                     PathBuf::from(p.to_str().unwrap().replace('/', "\\"))
                 } else {
                     p
-                };
-
-                // Jump to top-level project path to support passing paths
-                // from sub directories.
-                let top_level_path = config.src.join(&p);
-                if !config.src.join(&top_level_path).exists() {
-                    eprintln!("WARNING: '{}' does not exist.", top_level_path.display());
                 }
```

Comment:
Tested this locally with:

```
./x test test --exclude test
```

which reports: `Skipping Set({test::library/test}) because it is excluded`